mvebu: update u-boot and tf-a

dhewg opened the pull request:
This updates both components. Thanks to u-boot changes, we can cut the number of binary builds in half, since emmc is now detected at runtime.

~~Draft PR for now, since the tf-a bump creates a broken image.~~
~~tf-a bump isn't part of this PR anymore, it produces unbootable images~~

dhewg committed:
uboot-mvebu: update to v2021.01

u-boot now detects emmc variants at runtime, we don't need to build
seperate binaries anymore.

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg committed:
arm-trusted-firmware-mvebu: update to v2.4

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg committed:
arm-trusted-firmware-mvebu: don't build emmc variants

Starting with u-boot v2021.01 a single binary will be used for non-emmc
and emmc variants.

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg committed:
arm-trusted-firmware-mvebu: bump espressobin boards to CPU_1000_DDR_800

The cpufreq issue has been identified and a fix is in the process of beeing
upstreamed [0].

Bump the boards to the default 1000MHz so they can run at that frequency
once the fix is merged. Until then the boards are stuck at 800MHz (just
claiming to run 1000Hz, which is a lie).

[0] https://lore.kernel.org/linux-arm-kernel/20210114124032.12765-1-pali@kernel.org/

Signed-off-by: Andre Heider <a.heider@gmail.com>

dhewg commented:
And v2.4 also produces `uart-images.tgz.bin` now, the `.bin` suffix is new, that's not intentional, is it?

pali commented:
> And v2.4 also produces `uart-images.tgz.bin` now, the `.bin` suffix is new, that's not intentional, is it?

Seems that it was changed in following Marvell commit:
https://git.trustedfirmware.org/TF-A/trusted-firmware-a.git/commit/?id=fc8dc4999770dc0887b09e650f4ae7e7eb0befc0

So only filename/suffix was changed, not content.

pali commented:
> @pali with the atf bump to v2.4 the image boot loops, it doesn't even print the atf banner. It shrinks from 993336 to 862280 bytes, so probably I messed it up or it's just not including all components? Reverting the atf bump builds a bootable image again

I have not seen any building issue with the latest version from git master branch. Builds are working fine on both Espressobin v5 and Turris MOX. Can you try to build it from the git master branch?

Also are you sure that you are building according to steps https://trustedfirmware-a.readthedocs.io/en/latest/plat/marvell/armada/build.html ?

Have you compiled during OpenWRT prepare process both 32bit and 64bit gcc arm compilers?

If you do not figure out where is the problem, I will provide you in the next week full build command which I have used for building version for Espressobin v5 and also for MOX, which worked fine.

dhewg commented:
v2.3 release folder:
```
 12121 Jan 17 12:39 atf-ntim.txt
  1920 Jan 17 12:39 bl1
131072 Jan 17 12:39 bl1.bin
  1680 Jan 17 12:39 bl2
 16865 Jan 17 12:39 bl2.bin
  3480 Jan 17 12:39 bl31
 47712 Jan 17 12:39 bl31.bin
907320 Jan 17 12:39 boot-image.bin
907324 Jan 17 12:39 boot-image_h.bin
776248 Jan 17 12:39 fip.bin
993336 Jan 17 12:39 flash-image.bin
  1772 Jan 17 12:39 fuse.bin
    60 Jan 17 12:39 lib
   880 Jan 17 12:39 libc
    40 Jan 17 12:39 libwrapper
    40 Jan 17 12:39 romlib
  8464 Jan 17 12:39 sys_init.bin
  3324 Jan 17 12:39 TIM_ATF.bin
   120 Jan 17 12:39 uart-images
382661 Jan 17 12:39 uart-images.tgz
 10240 Jan 17 12:39 wtmi.bin
```

v2.4 release folder:
```
 12241 Jan 17 12:40 atf-ntim.txt
     0 Jan 17 12:40 bl1.bin
  1660 Jan 17 12:40 bl2
 16881 Jan 17 12:40 bl2.bin
  3460 Jan 17 12:40 bl31
 47712 Jan 17 12:40 bl31.bin
776264 Jan 17 12:40 boot-image.bin
776268 Jan 17 12:40 boot-image_h.bin
776264 Jan 17 12:40 fip.bin
862280 Jan 17 12:40 flash-image.bin
  1772 Jan 17 12:40 fuse.bin
    60 Jan 17 12:40 lib
   880 Jan 17 12:40 libc
    40 Jan 17 12:40 libwrapper
    40 Jan 17 12:40 romlib
  8672 Jan 17 12:40 sys_init.bin
  3324 Jan 17 12:40 TIM_ATF.bin
   120 Jan 17 12:40 uart-images
375389 Jan 17 12:40 uart-images.tgz.bin
 10448 Jan 17 12:40 wtmi.bin
```

I'm guessing it's because of `bl1.bin`

dhewg commented:
Ah, I switched the make targets from `all fip` to `mrvl_flash` like the doc you linked to reads. Using `all mrvl_flash` fixes `bl1.bin` again.

pali commented:
> Using `all mrvl_flash` fixes `bl1.bin` again.

This is strange... Because I remember that when I have tested build process, it worked fine also with just `mrvl_flash` target.

pali commented:
What version of WTMI image (from A3720-utils-marvell) are use using? From `A3700_utils-armada-18.12-fixed` branch or from the now new `master` branch?

pali commented:
And how you are building arm 32bit compiler?

dhewg commented:
My first test of a3700-utils were using 096797908ddd69a679fd55595c41fc02809829a9, unchanged from openwrt master. In the meantime I updated to 5598e150fa3a1568256c30223fd2b214d729f26a. Same issue though. I noticed that there's a 3.0 branch now, should we use that?

I also updated mv-ddr to 305d923e6bc4236cd3b902f6679b0aef9e5fa52d now.

The 32bit cross compiler is rather old binary toolchain downloaded from linaro: `5.0-2018.12-$(HOST_ARCH)_arm-linux-gnueabi`

pali commented:
New master branch of A3700-utils-marvell now points to the same commit as 3.0 branch which is the latest version. Same is for mv-ddr-marvell repository, new master branch points to the latest version.

Is there any change with updated mv-ddr-marvell?

That older binary 32bit arm toolchin should work. But is not OpenWRT compiling its own toolchain? Or it started using precompiled binaries?

dhewg commented:
openwrt builds a 64bit toolchain for kernel and userland, but not an additional one for cm3. Building that was controversial, so a binary one was used, which is also controversial, but that's how it is atm on master...

dhewg commented:
With the just pushed version I do get some BL1 output, but it still bootloops:
```
TIM-1.0
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltageNOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v2.4(release): (Marvell-devel-18.12.0)
NOTICE:  BL1: Built : 11:48:37, Jan 17 2021
NOTICE:  BL1: Booting BL2
NOTICE:  BL2: v2.4(release): (Marvell-devel-18.12.0)
NOTICE:  BL2: Built : 11:48:37, Jan 17 2021
NOTICE:  BL1: Booting BL31
NOTICE:  BL31: v2.4(release): (Marvell-devel-18.12.0)
NOTICE:  BL31: Built : 11:4WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
WTMI-devel-18.12.1-
WTMI: system early-init
CPU VDD voltage default value: 1.155V
...
```

See the 2.4 bump commit for the used versions.

pali commented:
You have something broken with WTMI build
```                                                                                                                                                                                                                                                                                                                                                                        
TIM-1.0                                                                                                                                                                                                                                                                                                                                                                    
WTMI-devel-18.12.1-                                                                                                                                                                                                                                                                                                                                                        
WTMI: system early-init
```
After the last dash in `WTMI-devel-18.12.1-` string should be version/git-commit of WTMI itself.

dhewg commented:
That was always the case as openwrt's package system downloads a tar of a specific commit, it doesn't build from a git clone directory.

pali commented:
You should not strip version information from builds. Then it makes hard for users to debug problems as they do not know what version are running.

pali commented:
`A3700-utils-marvell/wtmi/sys_init/Makefile` contains: `$(shell git log -n 1 --pretty=format:"%h")`

And `mv-ddr-marvell/scripts/localversion.sh` contains: `git -C $MV_DDR_ROOT rev-parse --verify --quiet --short HEAD 2> /dev/null` and `git -C $MV_DDR_ROOT diff-index --name-only HEAD 2> /dev/null`.
And then lot of other code which is also used to determinate if compiling from "official" Marvell tarballs available in Marvell Customer Extranet portal (under NDA) or from git.

So really, please do not strip these informations as it also modifies build process and cause problems...

dhewg commented:
I'm not stripping anything, the tool isn't able to identify its version. You'll get the same when building from a release tarball

pali commented:
But you are not building from release tarball, but from git. These tarballs differs from vesion in git.

pali commented:
You are really downloading it via git clone, but completely removing `.git` subdirectory:
https://github.com/openwrt/openwrt/blob/f4fdd8609c1e7c407956f1f65aa7f8d154af9875/include/download.mk#L214
This should not really be done for these wtmi/mv_ddr/tf-a repos.

dhewg commented:
I'm still not striping anything, what openwrt does it a completely established and reasonable approach.
And why're we even arguing about a freaking version string? Updating a component breaks booting, that sounds of more importance to me.

pali commented:
Well, what I'm trying to say, if you build a binary in _unsupported_ way and it does not work then people would look at it with less priority. And if you are compiling it in way that it strips all information which could help debug issue like what is the exact source code used for compilation then other people would not look at it deeply, as **all logs** are missing important information.

If such binary is working for you and you publish this binary to other people and they come up that it does not work, then it contains zero information how to debug issues which other people show. It just complicates all debugging if you remove the most important information from logs. And make people who want to investigate isses fall into assumption that binary was compiled from different source as logs claim.

If I'm going to debug this issue, the first thing which I need to know is version and source of build/code. I was debugging more these build issues of A3720 TF-A/WTMI/uboot and I know that these version strings are very important as more times I have been using incorrect binary (it was incorrectly compiled/re-compiled).

Therefore if you are going to strip version information from builds, I cannot help you with above mentioned issue as I really do not know if binary was correctly compiled and from which source.

I just do not want to debug again same issue that binary is not working just because was not recompiled or was incorrectly compiled. This stuff already take me lot of time.

dhewg commented:
Unsupported way? By whom? C'on, now your're making things up.

Anyway, of course a version string would be helpful. But that's not due to anyone stripping anything, it's a shortcoming of the tool and not the openwrt system that there's no way to provide that except by building from a full fledged git clone.

I dropped the atf bump as I don't have the time and energy to look into it. The whole build process is unfortunately still too fragile after all these years. But with this we now have an updated u-boot, half the binary builds and are prepared for the linux cpufreq fix!

pali commented:
I have explained it, the first step for debugging this issue is to show source/version from which was it compiled. I have seen more times that we have been debugging different source code of WTMI/TF-A as resulted binary and with this way, this version stripping just going to repeat same issues and same problems which were already solved.

From output which you sent it is even not known whether you have really built correct WMI binary and not just half-broken which happened more times.

So if you want to me to look at this issue, first compile resulted binary correctly including version strings and not strip some parts.

dhewg committed:
arm-trusted-firmware-mvebu: pass commit ids to a3700-utils/mv-ddr-marvell

The two required tools fail to identify their version when not compiling
from a git clone, patch that in and pass on the used commit hashes.

Upon boot it now prints "WTMI-devel-18.12.1-5598e150".

Signed-off-by: Andre Heider <a.heider@gmail.com>

pali commented:
> The whole build process is unfortunately still too fragile after all these years.

I fixed lot of issues in TF-A build process and added lot of checks in build process to prevent wrong things to happen. It should be less fragile. But never thought that somebody is going to remove some git files/directories during building boot image...

OpenWRT devs: are you aware of the fact providing firmware image on public website for end users with stripped version strings just makes resolving any issue impossible? Now there would be more images with same version strings (= empty string) and nobody would be able to tell which version of image is flashed on end user device.

More people wrote me emails during debugging CPU and other A3720 issues. I really cannot help to debug other A3720 issues if you want to provide just stripped firmware binaries without exact version information. Should I next time say to people who asking me about it just "do not use OpenWRT builds"?

dhewg committed:
trusted-firmware-a.mk: pass package version as version identifier

Upon boot it now prints:
NOTICE:  BL1: v2.4(release):OpenWRT v2.4-1 (espressobin-v3-v5-1gb-2cs) (Marvell-devel-18.12.0)

Signed-off-by: Andre Heider <a.heider@gmail.com>

DanielG commented:
Seems to me the wtmi Makefile really shouldn't rely on always being in a git checkout, but how then can it get the git version information, I hear you ask? Well, there is this neat git feature where you can tell git-archive to embed various bits of version information in certain files. What's more github supports this too, so any git source tarball downloaded using the API will also have this embedding happen.

See a quick example taken from one of my embedded projects here: https://gist.github.com/DanielG/d12035ea90eefd87d6ba1aad2ed2477f

Maybe this approach can be added to the upstream Makefile to fix this in the future?

dhewg commented:
Nice, I wasn't aware of `export-subst`.

But here the package itself is set to specific git hashes anyway, and the usual way is to just provide those through build or envvars to the project being compiled. You can already do that with atf with its make var `BUILD_STRING`, but not for the two used marvell  projects. That needs be patched in.

DanielG commented:
I belive our `dl_github_archive.py` script also has support for downloading tarballs by git commit or optionally repacking them from the git repo, so that was the operational mode I was thinking of. In both those cases the export-subst should then happen.

Keep in mind that for the GPL compliance use cases it is imperative to be able to build an openwrt image entirely offline. So then having tarballs in dl/ really is the only way.

hauke commented:
OpenWrt automatically creates tarballs from git repositories, this also makes the download fast compared to a git clone. 
We could try to add some meta information into the automatically generated tarball, but I would do this in a separate pull request.
How does it work when you use the official version downloaded from the Marvell website under NDA? Does this version has an extra file, we could also add such a file?

The trusted firmware a is not updated in this pull request. Are there any open Todos?

dhewg commented:
* added a patch so the tools accept an externally set commit id, which the build system then just uses
* added a patch so set atf's BUILD_STRING according to the openwrt package
* re-add atf bump, the issue is resolved with the master branch of both a3700-utils and mv-ddr-marvell

hauke commented on package/boot/arm-trusted-firmware-mvebu/Makefile:
```diff
@@ -32,7 +32,7 @@ define Trusted-Firmware-A/espressobin-512mb
   UBOOT:=espressobin
   DDR_TOPOLOGY:=0
-  CLOCKSPRESET:=CPU_800_DDR_800
+  CLOCKSPRESET:=CPU_1000_DDR_800
```
Will this update to 1GHz CPU clock work stable without the kernel changes?

dhewg replied:
Then it will actually never run at 1GHz. It claims it does, but it's stuck at 800MHz. I'm running debian with this patch (and its unpatched kernel) stable for weeks now.

It's unstable with an older kernel patch that just fixes it reaching 1GHz, without the AVS bump. The new kernel patchset fixes both, see the link in this patch.

hauke replied:
ok, so with current openwrt kernel it would just run at 800MHz anyway and there are some patches for mainline kernel which make it work stable with 1GHz and there is an older patchset which was not in OpenWrt or mainline which makes it unstable. Is this correct?

dhewg replied:
Correct. What I linked to is the proper patch set. The older patch was only used by armbian afaik.

hauke commented:
Thank you for the patch, I applied it to master.
